bin/download.tk1-exec-uboot

```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2015, NVIDIA CORPORATION. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+if [ -n "${board_usb_dev}" ]; then
```

Comment:
We already have download.tegra-uboot-flasher for this purpose; no need to duplicate it.  I guess this is related to your udev issue? Let's just fix that.

I wonder if the udev issue has anything to do with the rule being written for 4-10.1.1, yet your manual test command is for 4-10.1.2. There seems to be an inconsistency there.

Reply:
Yes sorry, that should be sudo udevadm test /sys/bus/usb/devices/4-10.1.1

But udev overall is not really working for me. Unless I 'udevadm test' 

Oddly udev does show the device appearing if I use 'systemctl status udev'

Feb 04 08:54:38 kea mtp-probe[3323]: checking bus 4, device 39: "/sys/devices/pci0000:00/0000:00:14.0/usb4/4-10/4-10.1/4-10.1.1"
Feb 04 08:54:38 kea mtp-probe[3323]: bus: 4, device: 39 was not an MTP device

If I use 'sudo udevadm test /sys/bus/usb/devices/4-10.1.1' while the test is waiting for /dev/usbdev-jetson-tk1 to appear, then device appears the test continues to run.

BTW on reboot, the machine only has a subset of the devices available, until I trigger them:

$ ./list.sh 
ls: cannot access '/dev/sdwire*': No such file or directory
ls: cannot access '/dev/ykush*': No such file or directory
ls: cannot access '/dev/usbdev-*': No such file or directory
ls: cannot access '/dev/em100*': No such file or directory
 /dev/relay0    /dev/sdcard2	    /dev/ttyusb_port2   /dev/ttyusb_port5   /dev/ttyusb_port8
 /dev/sdcard0   /dev/sdcard3	    /dev/ttyusb_port3   /dev/ttyusb_port6
 /dev/sdcard1   /dev/ttyusb_port1   /dev/ttyusb_port4   /dev/ttyusb_port7

I'm not sure if that subset has anything in common, though.

I'm pretty sure this used to work OK and not sure what I changed!

Reply:
On my system, when I run `udevadm info -a /dev/usbdev-jetson-tk1` I see a line containing `KERNEL=="1-8.3"` yet my udev rule contains `KERNELS=="1-8.3"`. Note the mismatch in the trailing `S`. I changed my rule to `KERNEL=="1-8.3"` and it still matches. I'm puzzled why it was working before given the requested match doesn't actually match!

Anyway, I wonder if `s/KERNELS/KERNEL/` in your rule would help at all? If not, what does the `udevadm info -a` output show for your device?

Reply:
`man udev` implies that KERNEL/KERNELS check the same attribute, but KERNEL checks the event device and KERNELs searches upwards, presumably starting in the same place, so this (a) explains why my original rule worked, and (b) likely means that changing your rule as I suggested won't fix anything.

Reply:
`udevadm test ...` injects a synthetic event for the device. If that does something but simply plugging in the device doesn't, I wonder if your kernel is actually generating udev events. Does `udevadm monitor` show anything when you unplug/replug in the device?

I have no idea why `systemctl status udev` would help...

Reply:
Actually 'systemctl status udev doesn't' help. It just shows that the device has been detected.

I can tell with 'udevadm monitor -e' that the kernel is seeing the device.

Reply:
With a bit more debugging:

sudo pkill udevd
sudo /lib/systemd/systemd-udevd -D

I see:

update old name, '/dev/usbdev-jetson-tk1' no longer belonging to '/devices/pci0000:00/0000:00:14.0/usb4/4-10/4-10.1/4-10.1.1'
no reference left, remove '/dev/usbdev-jetson-tk1'

and found this:

https://www.spinics.net/lists/hotplug/msg06086.html

So it seem that I need a condition like 'add|bind'.

Reply:
Oh, this is the systemd implementation of udev. I wonder if that's the reason or whether it's kernel version. Either way, I guess I'll need to re-write my rules during some future OS upgrade. Glad to see you found the issue; I'll review the updated patch.